<launch>
  <arg name="use_sim" default="false" />
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]" />
  <arg name="set_base_frame" default="base_footprint" />
  <arg name="set_odom_frame" default="odom" />
  <arg name="set_map_frame" default="map" />
  <arg name="maze" default="turtlebot" /> <!-- change to "final_maze" for a more open/"easy" map -->

  <!-- TurtleBot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch" />

  <!-- Sim specific -->
  <group if="$(arg use_sim)">
    <!-- Gazebo -->
    <include file="$(find lab4)/launch/world_sim.launch">
      <arg name="model" value="$(arg model)" />
      <arg name="world" value="$(arg maze)" />
    </include>

    <!-- Gmapping -->
    <node pkg="gmapping" type="slam_gmapping" name="turtlebot3_slam_gmapping">
      <param name="base_frame" value="$(arg set_base_frame)" />
      <param name="odom_frame" value="$(arg set_odom_frame)" />
      <param name="map_frame" value="$(arg set_map_frame)" />
      <rosparam command="load" file="$(find lab4)/config/gmapping_sim_params.yaml" />
    </node>

    <!-- Path Planner -->
    <node pkg="lab3" type="path_planner.py" name="path_planner" output="screen"
      args="--use_dynamic_map --padding 5 -n" />
  </group>
  <group unless="$(arg use_sim)">
    <!-- Gmapping -->
    <node pkg="gmapping" type="slam_gmapping" name="turtlebot3_slam_gmapping">
      <param name="base_frame" value="$(arg set_base_frame)" />
      <param name="odom_frame" value="$(arg set_odom_frame)" />
      <param name="map_frame" value="$(arg set_map_frame)" />
      <rosparam command="load" file="$(find lab4)/config/gmapping_params.yaml" />
    </node>

    <!-- Path Planner -->
    <node pkg="lab3" type="path_planner.py" name="path_planner" output="screen"
      args="--use_dynamic_map --padding 10 -n" />
  </group>

  <!-- Frontier Finder -->
  <node pkg="lab4" type="frontier_finder.py" name="frontier_finder" output="screen" />

  <!-- Maze Explorer -->
  <node pkg="lab4" type="maze_explorer.py" name="maze_explorer" output="screen" />

  <!-- Path Executor -->
  <node pkg="lab4" type="go_to_point.py" name="go_to_point" output="screen" />

  <!-- RViz -->
  <node pkg="rviz" type="rviz" name="rviz" required="true"
    args="-d $(find lab4)/rviz/explore_maze.rviz" />

</launch>